### PR TITLE
BlackoilModelEbos: separate out NLDD code to separate class

### PR DESCRIPTION
**Pull request by akva2**
More separation of concerns.

### CODE REVIEW

**Comment by akva2** [comment]
jenkins build this please

**Comment by akva2** [comment]
jenkins build this please

**Comment by akva2** [comment]
jenkins build this please

File: opm/simulators/flow/BlackoilModelEbosNldd.hpp [hunk 1]
```diff
@@ -0,0 +1,801 @@
...
+            ss << std::setw(4) << iteration;
+            for (int compIdx = 0; compIdx < numComp; ++compIdx) {
+                ss << std::setw(11) << mass_balance_residual[compIdx];
+            }
+            for (int compIdx = 0; compIdx < numComp; ++compIdx) {
+                ss << std::setw(11) << CNV[compIdx];
+            }
+            ss.precision(oprec);
+            ss.flags(oflags);
+            OpmLog::debug(ss.str());
```
Comment (bska) [inline opm/simulators/flow/BlackoilModelEbosNldd.hpp:605]: I don't see the requisite `<sstream>`, `<iomanip>`, and `<ios>` headers being included directly, so we're **possibly** depending on transitive include statements here.  We probably were in the original code as well, but now might be a good time to fix that omission.

File: opm/simulators/flow/BlackoilModelEbosNldd.hpp [hunk 2]
```diff
@@ -0,0 +1,801 @@
...
+                    for (const int c : domain.cells) {
+                        press_sum += solution[c][Indices::pressureSwitchIdx];
+                    }
+                    const double avgpress = press_sum / domain.cells.size();
+                    avgpress_per_domain[domain.index] = std::make_pair(avgpress, domain.index);
+                }
+                // Lexicographical sort by pressure, then index.
+                std::sort(avgpress_per_domain.begin(), avgpress_per_domain.end());
+                // Reverse since we want high-pressure regions solved first.
+                std::reverse(avgpress_per_domain.begin(), avgpress_per_domain.end());
```
Comment (bska) [inline opm/simulators/flow/BlackoilModelEbosNldd.hpp:651]: You strictly speaking need `<algorithm>` for `std::sort()` and `std::reverse()`.

File: opm/simulators/flow/BlackoilModelEbosNldd.hpp [hunk 3]
```diff
@@ -0,0 +1,801 @@
...
+                solveDomainGaussSeidel(solution, locally_solved, local_report,
+                                       iteration, timer, domain);
+                break;
+            }
+            // This should have updated the global matrix to be
+            // dR_i/du_j evaluated at new local solutions for
+            // i == j, at old solution for i != j.
+            if (!local_report.converged) {
+                // TODO: more proper treatment, including in parallel.
+                OpmLog::debug("Convergence failure in domain " + std::to_string(domain.index));
```
Comment (bska) [inline opm/simulators/flow/BlackoilModelEbosNldd.hpp:214]: You need `<string>` for `std::to_string()`.

File: opm/simulators/flow/BlackoilModelEbosNldd.hpp [hunk 4]
```diff
@@ -0,0 +1,801 @@
...
+                const int num_vars = residual[0].size();
+                std::vector<std::pair<double, int>> maxres_per_domain(domains_.size());
+                for (const auto& domain : domains_) {
+                    double maxres = 0.0;
+                    for (const int c : domain.cells) {
+                        for (int ii = 0; ii < num_vars; ++ii) {
+                            maxres = std::max(maxres, std::fabs(residual[c][ii]));
+                        }
+                    }
+                    maxres_per_domain[domain.index] = std::make_pair(maxres, domain.index);
```
Comment (bska) [inline opm/simulators/flow/BlackoilModelEbosNldd.hpp:669]: You need `<utility>` for `std::make_pair()`, and for `std::pair<>` for that matter.

File: opm/simulators/flow/BlackoilModelEbosNldd.hpp [hunk 5]
```diff
@@ -0,0 +1,801 @@
...
+        const auto& ebosResid = ebosSimulator.model().linearizer().residual();
+
+        for (const int cell_idx : domain.cells) {
+            const double pvValue = ebosProblem.referencePorosity(cell_idx, /*timeIdx=*/0) *
+                                   ebosModel.dofTotalVolume(cell_idx);
+            const auto& cellResidual = ebosResid[cell_idx];
+            bool cnvViolated = false;
+
+            for (unsigned eqIdx = 0; eqIdx < cellResidual.size(); ++eqIdx) {
+                using std::fabs;
```
Comment (bska) [inline opm/simulators/flow/BlackoilModelEbosNldd.hpp:780]: You need `<cmath>` for `std::fabs()`.

**Review by bska — COMMENTED** [review]
Thanks a lot.  Reducing the size of `BlackoilModelEbos.hpp` is very welcome.

I know you're "just" moving code around so it's not your fault, but it's annoying to see bare `size_t`'s being used without the requisite C header `<stddef.h>`.  Would you mind switching those to `std::size_t` from `<cstddef>` instead?

I also notice a couple of other missing headers from the standard library.  We were probably missing those before too, but now's a good time to fix that.

File: opm/simulators/flow/BlackoilModelEbosNldd.hpp (hunk 1 above)
Comment (akva2) [reply opm/simulators/flow/BlackoilModelEbosNldd.hpp:614]: actually they are in the blackoilmodelebos.hpp but i forgot to add them here. fixed.

File: opm/simulators/flow/BlackoilModelEbosNldd.hpp (hunk 2 above)
Comment (akva2) [reply opm/simulators/flow/BlackoilModelEbosNldd.hpp:660]: yeah, actually noticed and fixed locally but forgot to push.

File: opm/simulators/flow/BlackoilModelEbosNldd.hpp (hunk 3 above)
Comment (akva2) [reply opm/simulators/flow/BlackoilModelEbosNldd.hpp:223]: added.

File: opm/simulators/flow/BlackoilModelEbosNldd.hpp (hunk 4 above)
Comment (akva2) [reply opm/simulators/flow/BlackoilModelEbosNldd.hpp:678]: indeed, added.

File: opm/simulators/flow/BlackoilModelEbosNldd.hpp (hunk 5 above)
Comment (akva2) [reply opm/simulators/flow/BlackoilModelEbosNldd.hpp:789]: added.

**Comment by akva2** [comment]
and i have changed to std::size_t.

**Comment by akva2** [comment]
jenkins build this please

**Review by bska — APPROVED** [review]
Thanks for the updates.  This looks good to me now.  Please merge when the build check is green.

**Pull request merged**